Error messages extracted from golangci-lint output doesn't contain the true error information

<!--
Before filing an issue, please check if vim-go's help addresses your problem (see `:help go-troubleshooting`).

Consider executing `:GoReportGitHubIssue` to populate much of this information automatically.
-->
### What did you do? (required: The issue will be **closed** when not provided)
1. Open a Go file with errors, eg, containing unused variables:
```
package main                                                                                                                
                                                                                                                            
import "fmt"                                                                                                                
                                                                                                                            
func main() {                                                                                                               
  a := 1 // a declared but not used                                                                                                                    
  fmt.Println("Hello World")                                                                                                
}
```
2. Call `:GoMetaLinter`

<!--
If possible, please provide clear steps for reproducing the problem.
-->

### What did you expect to happen?
Shows error messages that contains `a declared but not used` information in the quickfix window.

### What happened instead?
Displays "Running error: asmdecl: analysis skipped: errors in package" in quickfix window

### Configuration (**MUST** fill this out):

#### vim-go version:
c074280f70cc838d113e0d293fabc0a0807cea85

####  `vimrc` you used to reproduce:
<!--
Use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc.

If this is not provided or is obviously incomplete, the issue may be unceremoniously closed.
-->
<!-- vimrc -->
<details><summary>vimrc</summary><br><pre>
let g:go_def_mode='gopls'
let g:go_info_mode='gopls'
let g:go_fmt_autosave = 1 
let g:go_fmt_command = 'goimports'
let g:go_addtags_transform = 'snake_case'
let g:go_highlight_types = 1 
let g:go_highlight_fields = 1 
let g:go_highlight_function_calls = 1 
let g:go_highlight_operators = 1 
let g:go_list_height = 10
let g:go_list_type = "quickfix"
let g:go_list_autoclose = 1 
let g:go_metalinter_enabled = ['vet', 'golint', 'errcheck']
let g:go_metalinter_autosave = 1 
let g:go_metalinter_autosave_enabled = ['vet', 'golint', 'errcheck']
</pre></details>

#### Vim version (first three lines from `:version`):
<!-- :version -->
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 18 2020 18:29:15)
Included patches: 1-1453
Modified by pkg-vim-maintainers@lists.alioth.debian.org
```

#### Go version (`go version`):
<!-- go version -->
`go1.13.9 linux/amd64`

#### Go environment
<details><summary><code>go env</code> Output:</summary><br><pre>
<!-- go env -->
GO111MODULE=""
GOARCH="amd64"
GOBIN="/home/dz/work/bin"
GOCACHE="/home/dz/.cache/go-build"
GOENV="/home/dz/.config/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/dz/work"
GOPRIVATE=""
GOPROXY="https://goproxy.io,direct"
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build024599614=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

#### gopls version
<details><summary><code>gopls version</code> Output:</summary><br><pre>
<!-- gopls version -->
golang.org/x/tools/gopls v0.3.4
    golang.org/x/tools/gopls@v0.3.4 h1:4GC7q/pXQ/tsxHBGVdsMdlB4gCxVC06m/7rIXg1Px4E=
</pre></details>

#### Golangci-lint version
`golangci-lint has version 1.24.0 built from 6fd4383 on 2020-03-15T11:38:02Z`

### Reason

The golangci-lint tool's output for unused variable is (with infos and warnings omitted):
```
level=error msg="Running error: asmdecl: analysis skipped: errors in package: [/path/to/main.go:6:2: a declared but not used]"
```
But the `errorformat` for golangci-lint output is
```
level=%trror\ msg="%m:\ [%f:%l:%c:\ %.%#]"
```
which extracts _Running error: asmdecl: analysis skipped: errors in package_ by `%m` but ignores _a declared but not used_ by `%.%#`

### Quick fix

Here's my [quick fix](https://github.com/lucas884e/vim-go/commit/0081dcf558d28a077f361495e836fd58105882be#diff-def140e42e54c1997c76370fbca46c87R446)

(I'm not a vimscript expert nor familiar with vim-go codebase. So forgive me, I'm not going to open a pull request)

--- a/autoload/go/lint.vim
+++ b/autoload/go/lint.vim
@@ -442,14 +442,40 @@ endfunction
 
 function! s:errorformat(metalinter) abort
   if a:metalinter == 'golangci-lint'
-    " Golangci-lint can output the following:
-    "   <file>:<line>:<column>: <message> (<linter>)
-    " This can be defined by the following errorformat:
-    return 'level=%tarning\ msg="%m:\ [%f:%l:%c:\ %.%#]",level=%tarning\ msg="%m",level=%trror\ msg="%m:\ [%f:%l:%c:\ %.%#]",level=%trror\ msg="%m",%f:%l:%c:\ %m,%f:%l\ %m'
+    let l:efm = ''
+
+    " Golangci-lint can several formats, all of which seem to be undocumented.
+    " Based on trial and error, these error format strings seem to catch all
+    " the relevant combinations.
+
+    " When the actual locations and error message is wrapped in parentheses
+    " within brackets (there is usually duplicated location information in
+    " this form). This usually happens when the linter can't do its job
+    " because of compiler or AST problems.
+    let l:efm .= 'level=%tarning\ msg="%.%#(%f:%l:%c:\ %m)\ %.%#]"'
+    let l:efm .= ',level=%trror\ msg="%.%#(%f:%l:%c:\ %m)\ %.%#]"'
+
+    " When the actual locations and error message are not wrapped in
+    " in bracked without the inner parenthetical. This usually happens when
+    " the linter can't do its job because of compiler or AST problems.
+    let l:efm .= ',level=%tarning\ msg="%.%#:\ [%f:%l:%c:\ %m]"'
+    let l:efm .= ',level=%trror\ msg="%.%#:\ [%f:%l:%c:\ %m]"'
+
+    " When the file location is not provided. The usually happens when the
+    " linter can't do its job because of some other problem.
+    let l:efm .= ',level=%tarning\ msg="%m"'
+    let l:efm .= ',level=%trror\ msg="%m"'
+
+    " when the linter was able to compile and/or create the AST, and the
+    " linter found some problems.
+    let l:efm .= ',%f:%l:%c:\ %m'
+    let l:efm .= ',%f:%l\ %m'
   elseif a:metalinter == 'gopls'
-    return '%f:%l:%c:%t:\ %m,%f:%l:%c::\ %m'
+    let l:efm = '%f:%l:%c:%t:\ %m'
+    let l:efm .= ',%f:%l:%c::\ %m'
   endif
 
+  return l:efm
 endfunction
 
 function! s:preserveerrors(autosave, listtype) abort

--- a/autoload/go/lint_test.vim
+++ b/autoload/go/lint_test.vim
@@ -53,10 +53,9 @@ func! s:gometa_problems(metalinter) abort
   try
     let g:go_metalinter_command = a:metalinter
     let expected = [
-          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': '[runner] Can''t run linter golint: golint: analysis skipped: errors in package'},
-          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'Running error: golint: analysis skipped: errors in package'}
+          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': 'import \"/quux\": cannot import absolute path'},
+          \ {'lnum': 3, 'bufnr': bufnr('%'), 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'import \"/quux\": cannot import absolute path'},
         \ ]
-
     " clear the quickfix lists
     call setqflist([], 'r')
 
@@ -141,8 +140,8 @@ func! s:gometaautosave_problems(metalinter) abort
   try
     let g:go_metalinter_command = a:metalinter
     let expected = [
-          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': '[runner] Can''t run linter golint: golint: analysis skipped: errors in package'},
-          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'Running error: golint: analysis skipped: errors in package'}
+          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'w', 'module': '', 'text': 'import \"/quux\": cannot import absolute path'},
+          \ {'lnum': 3, 'bufnr': bufnr('%')+1, 'col': 8, 'pattern': '', 'valid': 1, 'vcol': 0, 'nr': -1, 'type': 'e', 'module': '', 'text': 'import \"/quux\": cannot import absolute path'},
         \ ]
 
     " clear the location lists

